### DISCUSSION TITLE
[Design/General] Power Rework

### DISCUSSION THREAD

**Discussion by Sirryan2002** [ParadiseSS13/Paradise#19786] [Documents]
# Power Rework
## Introduction
Before power can seriously be tied into another system, much work must be done. Mainly a lot of work needs to be done to bring up-to-date old code and to tie all power machinery into a more holistic energy system. This rework is the first stepping stone in #19734

I really want to start bringing other contributors into these projects as they get larger. Below I've included a roadmap with some PRs that contributors can work on.

## The Current Powernet
![OldPowerSystem](https://user-images.githubusercontent.com/80364400/204114899-73e01e38-e7d7-4ef6-83f6-e78ffd4addde.png)

**How the current powernet works:**

- At round start, Power (Measured in Watts) is produced either through an engine or solar panels. At each of these energy producers, energy is transferred directly to a SMES
- A SMES is a storage unit with a base storage amount of 200MW. It can take in, store, and output power from one powernet into another (or back into its own powernet).
- Power is transferred along basic red wires either as a connection between power producers and SMES’s or from the SMES’s to various locations across the station
- APCs receive power through a terminal connected to these red wires and distribute that power to any machines in the APCs area. These APCs also hold power in reserve in case it receives no power from the powernet.
- 

**The problems with the current powernet:**

- (1) Power values aren’t balanced for demand to meet supply, so coders have compensated by guaranteeing that the roundstart engine and solar can power the station indefinitely as long as they’re turned on. This is an issue because it basically means that the power isn’t there except when an external force messes with it even though everything on the station uses it for the most part.
- (2) There is only one type of wire and every APC on the station must be a part of a net of these wires which eventually reaches the engine SMES’s. This results in what I’m calling a “spaghetti net”:
   - Engineers/Antagonists have no fucking clue what parts of the station wire net are “critical connections” i.e. the only points connecting large swathes of APCs to the engine. Even if they did know which of these wires are critical connections, there’s nothing about the wire which indicates that's what it is.
   - The powernet is a global net, if there’s a problem with the net, it affects every single APC. If the powernet is overloaded, every APC arcs; if there’s low power or a power sink, every APC is drained.
   - The common issue here is that issues are too large for engineers to wrap their head around, if an engineering player wants to have fun it will be gained by putting the problem right in front of their faces so they can try and solve it, not go on a “find the hidden object” quest. 
- (3) The powernet is complex in weird places and too simple where it matters:
   - Why give solars power production rates with a sin function when you slap a SMES in the mix which sheds all excess power and turns it essentially into a constant production rate???
   - Even though APCs have all these bells and whistles, they’re absolutely useless for diagnosing issues with the powernet. I love the art in its complexity but trying to navigate the breakers, wire panel, wire terminal to figure out why there’s no charge only to figure out its because theres a cut wire elsewhere is awful.
   - Power problems are a stupid quick fix such as replacing a single wire, finding a power sink, or pressing a button on the SMES, players wont even really use the game mechanic skills that an “electrician role” would be expected to know and master. This is just a violation of basic game design principles IMO.
- (4) Power consumption drops as machines are upgraded causing positive feedback loops. This one doesn’t really need much elaboration, increasing machinery production with upgraded components should come at a cost instead of pure benefit.
- (5) Unlike atmos, there’s not really any challenge to working with electricity except learning some obfuscated mechanics. There’s not some hurdle that players have to overcome which is fun to vault over; I’m basically calling our power mechanics dry.

## Proposed Changes
![ProposedPowerSystem](https://user-images.githubusercontent.com/80364400/204114895-1e22aff8-d6da-4414-90f7-481633c72783.png)

**How this powernet works:**

- Power is now differentiated by voltage, to keep things simple (looking at you atmos) power will just be in two states: High voltage or low voltage. High voltage can only travel through HV cable and Low Voltage can only travel through LV cable. The difference here is that HV is highly efficient and LV is inefficient so more and more power will be lost the longer the LV cable line.
- HV has to be stepped down to LV through transformers, next to these transformers will be breaker boxes which act as remote breakers for each APC. These will be situated in a room somewhere not locked behind departmental access and restricted to only engineers.
- Power producers make HV electricity which can be stored in SMES’s. Instead of going Engine Net -> SMES -> Station Net; Production is hooked directly into consumption and excess power needs to be stored or shed in some way
- Some form of power shedding method such as a tesla coil bluespace thingamabob that “deletes” energy but becomes increasingly more unstable the more energy you shed. This would force you to shut down other engines such as solar or dial back a supermatter/tesla/singulo to accommodate other fun engines.
- There is a “global” HV net that goes across the station and connects to “regional” or departmental LV nets. The main supply line will be very well defined and simple which goes to solve issues (2) and a bit of (3). Power issues are mostly quarantined to a regional net and so diagnosing the issue is more clear.

**Changes That need to be made to Engines:**

- **Solars:** Solars should no longer have a constant power rate. Instead all of them should have the same rate at the same time. This means that as the power rates go up and down with SSsun (of course this is a subsystem), every single solar panel is producing the same amount of power at the same time. This is so that the actual downside to free infinite power is that it’s not constant (its a sin wave function for rates)
- **PACMANs:** I’d like to keep these mostly the same, they’re pretty good. They just need their values adjusted. I would like PACMANs to be solely LV so they can serve their niche much better.
- **Singulo/Tesla:** These engines are too basic, they even give power or they’re destroying the entire station. There should be ways to mess with these engines in a way in which one can trade off risk for power, the singularity should be more than Bigger == More Power as well.
- **Supermatter:** Make it take up less space and become a much more adjustable engine, types of gas and emitter usage shouldn't be the only variables involved in adjusting power production of the SM. Will probably need a decent rework
- **TEG:** Very unbalanced at the moment, they need to have power production nerfed or have decreasing marginal returns past a certain point
- **Turbine:** Pretty unused but otherwise kinda balanced, mostly needs values adjusted.
- **NEW ENGINES:** I would like to see a form of a static constructed engine that produced power that is reactive to the demand on the network, much like steam engines in Factorio. Also some nuclear reactor or something, but it should probably wait until power is balanced.

## Other Proposed Additions:

- APCs give error messages based on why they’re not working. Wire panel cuts, lack of external power, etc.
- Overloaded APCs/Powernets no longer just arc. Instead, warning alarms are sounded form the APC and shit starts going down. After a certain amount of time, APCs arch, lights start brightening and popping, if it goes on too long, rare chance that a machine fries and eventually the APC shorts or the battery explodes.
- Still trying to brainstorm ways to make it so this only happens in one area first then the other so we don’t have global net issues immediately
- After a certain amount of time the breakers will just trip which will stop destruction from occurring but that net will no longer consume power leading possible to a cascading blackout caused by oversupply and falling demand. THIS GIVES ENGINEERING TIME TO FIX IT
- Small department battery SMES reserves, significant nerf of APC batteries
- Transformers can have their circuit “resistance” adjusted so that it changes the amount of watts it lets into the net

## Roadmap
This is where other people can help, each of these are their own PRs
⚠️ PRs I need to do myself    🤗 PRs any contributor can do  ✋ PRs that someone else can do but need to wait 

- [ ] #19991 APC and Powernet Refactor (big PR) ⚠️
- [ ] #20598 Powernet/Wirenet & Power Machine Refactor ⚠️
- [ ] #21250 HV Power & Transformers ⚠️
- [ ] #21250 Power Shedding, and Power Value Rebalance ⚠️
- [ ] #21250  Breaker Boxes ✋ (Needs to wait for transformers and local LV nets)
- [ ] #21250 APC Overload Rework ✋ (Needs to wait for transformers and local LV nets)
- [ ] #21250 Solars Rework & Code Refactor 🤗
- [ ] #19999 Tesla/Singulo rebalancing or rework (use your imagination) & refactor 🤗

**Comment by S34NW** [comment]
The breakerbox concept is really cool, what are the chances of things like diodes being added for one-way power transmission? (I suppose step down transforers will behave somewhat like this?)

Ironically, Cere station will work really well with these changes as we can run HV power to the substations already in place and step them down there so each asteroid is on it's own LV net.

**Reply by Sirryan2002** [reply]
Yes, so transformers will only either be step up or step down. This just means that they transform power in one direction either HV -> LV or LV -> HV. As for actual cable logic outside of transformers, I'm leaving that for the power module in progression trees as its a feature I want to build into those systems.

**Comment by Ciesson** [comment]
RE Power Monitors: In addition to the current Total Supply/Total Draw and APC listing on a power monitor, is there a gameplay related reason (to allow/enforce certain gameplay mechanics?) that we can't have an enumerated list of SMESes connected to the grid (that the power monitor is connected to), with their Input and/or Output? i.e.
```
SMES Status
---
Outputting to connected grid:
Engine SMES 1 - 100kW
Engine SMES 2 - 100kW
...
Aft Port Solar SMES - 60kW
Fore Port Solar SMES - 60kW
...
Secondary Electrical SMES 1 - 50kW
Secondary Electrical SMES 2 - 50kW
...
---
Inputting from connected grid:
Gravgen SMES - xxkW
Secondary Electrical SMES 1 - 100kW
Secondary Electrical SMES 2 - 100kW
---
```
Use cases:
- sanely managing "power backup loops" like the secondary power room (I know this specific setup may be going away in the rework, but would add immediate gains)
- with a bit of headwork, positively identifying the existence of a power sink on a grid with multiple SMES inputs/outputs
- a general improvement in monitoring the general "health" of the connected grid of a power monitor (give me another good and rewarding reason to sit in the control room and monitor the grid as an Electrician)

User story:
A lot of engineers in the last few rounds I have played have been experimenting with various exotic engines designs (which is awesome), often leading to brief or prolonged power outages across the station (I also approve).
However, a specific round that comes to mind had everyone running around like headless chickens looking for a supposed "power sink" at the same time as trying to save and restart a CO2 engine gone bad. It turned out to have been an adventurous engineer who setup the "backup power loop" in secondary electricals badly during early round, which worked fine when he set it up with loads of power available. As soon as the engine had to be shutdown, the grid then became unstable for the rest of the round, as it sucked up all the solar we were still producing. followed by secondary electrical getting infested with xenos and the equipment destroyed, with all that stored energy being lost. I got grabbed by a xeno on the way to investigate my theory in secondary electrical, ending my round.

I'm looking to get into this codebase and contribute to Paracode, and I figure something like this may be a good start, if it is desired?

I can't find any mention of improving the power monitor display in the GitHub discussions (but I am likely not checking all the correct places). As far as I can tell, with the first big power refactor merged, it should be possible to implement this now in a way that is forwards-compatible with upcoming changes.

**Reply by Sirryan2002** [reply]
Yeah, it should be easy enough to add SMES to the grid power monitoring computer. I would be in favor of it since that's the route I'm pretty much going to be taking with the next power overhaul PR coming up (transforms, LV/HV Power, Breaker Boxes, etc)
